### ISSUE TITLE
Replace Viewer Framework with IPython Widgets

### ISSUE THREAD

**Issue by blink1073** [scikit-image/scikit-image#1311]
See demo from [scikit-spectra](http://youtu.be/RhMHgQbP__A).
https://github.com/hugadams/scikit-spectra/blob/master/skspec/interact/ipynbs/specgui.py
Wiki for discussion [here](https://github.com/scikit-image/scikit-image/wiki/Web-based-Viewer-Proposal).

**Comment by blink1073** [comment]
@tacaswell here's a nice screenshot of my up-and-coming rectangle selector in the notebook:

![screenshot from 2015-01-02 10 31 28](https://cloud.githubusercontent.com/assets/2096628/5597603/b77b0e72-926a-11e4-8122-4c36347c27ba.png)

**Comment by tacaswell** [comment]
Cool!

Doing a final review of that PR is on my to-do list for (hopefully) today.

**Comment by JDWarner** [comment]
If anyone else wants to follow along, the PR by @blink1073 is matplotlib/matplotlib#3937

**Comment by blink1073** [comment]
Thanks @JDWarner.  Perhaps we could make the new viewer depend on `Matplotlib >= 1.4`, and use a local copy of `widgets.py` that has all the new widgets (Rectangle, Ellipse, Paint, Line, Lasso) until such a time as all of the new widgets are in a Matplotlib release, then switch over to Matplotlib's `widgets.py`?

**Comment by JDWarner** [comment]
@blink1073 :+1: assuming the only thing missing is the widgets, and not other necessary improvements (e.g., to the `nbagg` backend).

That's exactly what we did with `pad` to maintain backwards compatibility with NumPy 1.6 while using a NumPy 1.8+ function throughout the package. Great minds...?

**Comment by blink1073** [comment]
Shucks, we're on hold pending matplotlib/#3963.  `key_press_event` is not being handled by `nbagg` (but works in `webagg`).   I'd switch to `webagg`, but then we'd lose the IPython widgets...

**Comment by hugadams** [comment]
Thanks for CCing me on this thread, Steven.  Really a big fan of this idea.  Here are some issues that come to mind when deciding to use the IPython widget framework:

1.  The widget API is in a state of flux, and may eventually be taken out of IPython's repo completely and put in a separate repo.  The API is likely to change in the middle of your dev cycle, and it can be a pain to keep up.  I'd recommend establishing an IPython clone that everyone developing for the GUI uses during dev.

2.  Linking widgets to traits takes more boiler plate than building a GUI in TraitsUI, so there's a bit of a learning curve.  I hope that future versions if IPython will simplify the way `link` is used to connect widgets to traits.

3.  IPython uses Traitlets, not the full Traits library, so some important traits are not available.  For example, a useful widget might be a list that stores images in a `TableEditor`(http://code.enthought.com/projects/traits/docs/html/TUIUG/factories_advanced_extra.html#tableeditor), so users could go back and forth easily with a selected image.  IIRC, `List` traits are not part of Traitlets, and I'm sure the TableEditor isn't.  Most likely, Traitlets will limit the complexity of widgets and editors.  Offhand, I wonder if you could use full Traits on the Model side instead of Traitlets, or is Traitlets baked into the widgets?  This would add Traits as a dependency, but well worth it IMO.

4.  When desiging the IPython notebook GUI, we found that keeping separate Model and View classes was not too hard.  This is reflected in our classes `specmodel.Specmodel` and `specgui.SpectralGui`.  Feel free to pick through the source of these classes if they're helpful.  https://github.com/hugadams/scikit-spectra/tree/master/skspec/interact/ipynbs  Additionally, Nicholas Bollweg from the IPython core team helped us a great deal to get this up and running.  He is a whiz on designing widget GUIs and I'd recommend sending 'hard' questions to him.

5.  The ImageViewer, be it in QT, NBWidgets etc..., would probably benefit from binaries.  I think the reason ImageJ is so popular is because of how out-of-the box it's applet is.  Whether one frontend framework or another would be more suited to making executables might be an important factor to consider now.

The new matplotlib interactivity look pretty awesome.  Anyone know if there's a tool to select points in a 2d contour plot yet?  Specifically to get the z value at various x,y points?  Chaco is potentially another option, since it's already traits-aware and has many of these tools builtin.  I just shared a demo that showed Chaco and scikit-image working together, but it's certainly not as well-known as matplotlib.  With the pace that interactive matplotlib is going, it's probably superfluous.  Just like I did in the demo, maybe the viewer would benefit by supported multiple plotting backends, MPLD3, Chaco and Matplotlib if possible, with Matplotlib being the primary library, and switching is only recommended when a certain tool is not yet supported.  Or maybe this would clog up dev too much to be worthwhile.

I'll try to keep an eye on this thread.  We have an undergrad working on the next scikit-spectra GUI, and I'm bringing in my colleague, @EvelynLiu77, who did most of the work on the notebook GUI and might be able to help contribute to this project.

PS, early on I saw bokeh as a dependency for mpld3?  It should not be, or has this been added recently?

Thanks

**Comment by hugadams** [comment]
Actually, chaco won't port easily into the notebook AFAIK, so ignore all my previous mentions of it.  Getting late..

**Comment by blink1073** [comment]
@hugadams, thanks for joining the conversation!

Our goal is to reduce *our* dependencies, hence the move away from Qt.  Bokeh is not a dependency for mpld3, I meant that Bokeh would be a large dependency for us.  We already have matplotlib as a dependency, and I've been working on improving their webagg/nbagg event handling, which will be in their 1.4.3. release.  

See [datacursor](https://github.com/joferkington/mpldatacursor) for your contour picker question.

I agree that trading Qt for a moving target (IPython) is not necessarily ideal, but we've seen that requiring Qt can be a barrier to entry.  

TraitsUI is out because they do not yet support Python 3 (and the core devs have shown no indication that they are planning to any time soon).

We could also hand-roll our own UI using JQueryUI and websockets.  I think I can replicate what we have so far using that approach, but future growth would be limited by our lack of JS expertise.

Weighing all of the above, I think that Bokeh's growing stable of widgets and tools is attractive, and we would benefit from their (stable) growth.  As much as I like Matplotlib, I don't like being tied to the notebook and its changing interface.

**Comment by stefanv** [comment]
@blink1073 Do you think this is something we can feasibly attack at the moment?  If so, perhaps a GSoC is indicated.

**Comment by blink1073** [comment]
@stefanv, I would not be opposed to the idea.

**Comment by stefanv** [comment]
@scikit-image/core How would you feel about building our widget sets on top of Bokeh?

**Comment by blink1073** [comment]
FWIW I tried `pip install bokeh` on my 14.04 system python (2.7), and it built with all dependencies smoothly.  All the dependencies are also available on Christoph's site for folks without a sane build system (and aren't using Anaconda).

**Comment by matthew-brett** [comment]
No Bokeh wheels for OSX though.

How widely used are Bokeh widgets?

**Comment by blink1073** [comment]
Wait, people still use OSX?   :smile:   Hard to say how widely used the widgets are, they were introduced last summer.  They also have IPython notebook integration, and their linked data table is quite lovely.  http://continuum.io/blog/bokeh-0.7 has a few video clips.

**Comment by hugadams** [comment]
I wonder why IPython developed their own separate widget system

On Wed, Feb 4, 2015 at 10:46 PM, Steven Silvester <notifications@github.com>
wrote:

> Wait, people still use OSX? [image: :smile:] Hard to say how widely used
> the widgets are, they were introduced last summer. They also have IPython
> notebook integration, and their linked data table is quite lovely.
> http://continuum.io/blog/bokeh-0.7 has a few video clips.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/issues/1311#issuecomment-72989527>
> .
>



-- 
Adam Hughes
Physics Ph.D Candidate
George Washington University

**Comment by JDWarner** [comment]
@blink1073 Still using the glory days of OSX 10.6.8... Anaconda works perfectly, so I've never tried out wheels.

**Comment by jni** [comment]
@matthew-brett I'm with @JDWarner, who wants wheels when there's Anaconda?

I'm of course *not* with @blink1073 in that I am using OSX (10.10.2) and loving it. =D

Bokeh is still in its infancy, and I don't imagine that the API is very stable. However, that is also true of IPython widgets, and Bokeh has the explicit goal of being highly performant for big datasets, which I like. I generally *love* everything that Continuum are doing, so I'd be :+1: for choosing Bokeh.

Having said this, if I see another "from bokeh.plotting import *", I will not be held accountable for my actions.

**Comment by matthew-brett** [comment]
Probably not worth going into wheels vs Anaconda, except to say that I would rather we (scientific Pythonistas) do not require people to install Anaconda to get work done.   You can currently get the scipy stack and various other big packages with wheels using a default Python.org or homebrew or macports install.

I would say Bokeh is a heavy dependency, but I guess that would be OK if it was fully optional.

**Comment by stefanv** [comment]
Would it be hard to build wheels for Bokeh? I like to hedge my bets.

**Comment by matthew-brett** [comment]
See: https://github.com/bokeh/bokeh/blob/master/setup.py#L398

Very long requirements list.  Binary requirements then need wheels.  So,
would require some legwork to make wheel-building travis repo + persuade
maintainers of gevent in particuar to use this or to provide wheels out of
the box.    If you wait a while I suspect the maintainers will get round to
it, as wheels become standard practice for OSX.

**Comment by stefanv** [comment]
Here's the list:

```
'Flask>=0.10.1', 'Jinja2>=2.7', 'MarkupSafe>=0.18', 
'Werkzeug>=0.9.1', 'greenlet>=0.4.1', 'itsdangerous>=0.21', 
'python-dateutil>=2.1', 'pytz==2013b', 'requests>=1.2.3', 
'six>=1.5.2', 'pygments>=1.6', 'pystache>=0.5.3', 
'markdown>=2.3.1', 'PyYAML>=3.10', 'pyzmq>=14.3.1', 
'tornado>=4.0.1', 'click>=3.3', 'colorama>=0.2.7'
```

Of those, perhaps only two or three are binary requirements (and I think
those are requirements for IPython as well).

**Comment by matthew-brett** [comment]
pyzmq has binary wheels (hence IPython notebook is an easy dependency).

I believe Markupsafe and PyYaml and tornado have fall-back slow pure Python implementations when the compile step doesn't work.  I'm guessing that at least PyYaml is quite a bit slower without compiling, but whether that would be a problem in practice I don't know.

Further down the setup.py file:

```
if sys.version_info[0] != 3 and platform.python_implementation() != "PyPy":
    REQUIRES.extend([
        'websocket>=0.2.1',
        'gevent>=1.0',
        'gevent-websocket>=0.9.2',
    ])
```

which is where the `gevent` dependency comes from.  This seems to be a considerable binary dependency.

**Comment by matthew-brett** [comment]
Oh - and I think you will also get Pandas as a dependency:

```
if platform.python_implementation() != "PyPy":
    # You need to install PyPy's fork of NumPy to make it work:
    # pip install git+https://bitbucket.org/pypy/numpy.git
    # Also pandas is not yet working with PyPy .
    REQUIRES.extend([
        'numpy>=1.7.1',
        'pandas>=0.11.0'
    ])
```

**Comment by stefanv** [comment]
That is quite heavy, indeed.  It almost sounds like the playing field is
not very ready for what we want to do here.  Perhaps we can consider a very
simple JavaScript canvas of our own without any dependencies?

**Comment by hugadams** [comment]
What are the main drawbacks of the current QT setup?

On Thu, Feb 5, 2015 at 1:52 PM, Stefan van der Walt <
notifications@github.com> wrote:

> That is quite heavy, indeed. It almost sounds like the playing field is
> not very ready for what we want to do here. Perhaps we can consider a very
> simple JavaScript canvas of our own without any dependencies?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/issues/1311#issuecomment-73103502>
> .
>



-- 
Adam Hughes
Physics Ph.D Candidate
George Washington University

**Comment by blink1073** [comment]
@hugadams .  For one, we have to support three versions: PyQt4/5 and PySide.  Also, it is difficult to get set up, especially in a virtual environment: see our Travis [script](https://github.com/scikit-image/scikit-image/blob/master/tools/travis_script.sh#L18).   Finally, I personally see browsers as the future of user interfaces. 

@stefanv, like I said, I'm game to roll a simple JQueryUI setup.

**Comment by matthew-brett** [comment]
A simple setup would be a great thing :)

**Comment by jni** [comment]
@matthew-brett you're absolutely right re: not requiring Anaconda, of course. =)

Like @blink1073, I think cross-platform UIs definitely belong in browsers.

**Comment by stefanv** [comment]
JQueryUI would give us some buttons, but I am more concerned about 
the actual canvas tools.  And, then, there is the question of 
testing (I know how to test JavaScript utility libraries, but I 
have no experience testing interactive widgets).  Does anyone 
else?

Strategically, is this a good way to spread our resources?  I 
would argue "yes" if a) it adds significant benefit to the package 
and b) it is an enjoyable endeavour.  But "no" if it's better to 
punt and let others (like Bokeh) develop these tools.

/cc @takluyver

**Comment by blink1073** [comment]
@stefanv, the canvas tools are still just matplotlib tools, event handling is fully supported in webagg/nbagg as of mpl v1.4.3.

**Comment by stefanv** [comment]
You'll have to educate me: does the webagg backend run a
matplotlib-specific server, or how does it connect the JS frontend to Python?

**Comment by JDWarner** [comment]
A couple thoughts:

We definitely should get decent UI functionality out there, but I don't know if spreading our resources across different languages is worth the benefit. It could be; these user-friendly interfaces have obvious utility. On the other hand, a flood of new (and possibly less technically inclined) users could result in high burden of upkeep and support. Which would be 100% on us, if we roll our own.

Personally I feel the *NIX philosophy of "do one thing, and do it well" has served us well thus far. With no clear path ahead, another option would be to closely observe from the sidelines while these options mature toward stability, and revisit the issue in a couple months. Perhaps around the SciPy 2015 sprints ;)

**Comment by blink1073** [comment]
@stefanv, we'd launch a tornado web app, create an mpl figure in python and JS, and the canvas events are passed between the two.  We'd have our own websocket to share state between python and JS. (Basically I'd copy what they're doing and add another JS script with our pieces).

@JDWarner, the intent would be for there to be as little JS code as possible (just some sliders and buttons with callbacks).  @tacaswell and I came up to speed on JS pretty quick and were able to hash out the canvas event handling behaviour, which I'd argue is trickier than anything we'd have to do here.

**Comment by jni** [comment]
I'm inclined to agree with @JDWarner here: is there even any reason for this viewer to be in skimage proper, rather than a separate repo, e.g. skimage-viewer?

**Comment by blink1073** [comment]
Don't look at me, blame @tonysyu!

**Comment by tonysyu** [comment]
Actually, my first cut was a completely separate package. @stefanv actually convinced me to move it into skimage :D

**Comment by jni** [comment]
LOL!!!

But we could even:

- deprecate the viewer subpackage
- move it to a separate repo, `skimage-qt-viewer`
- create other repos as needed, e.g. `skimage-ipython-widgets`, `skimage-bokeh-widgets`, etc.

**Comment by blink1073** [comment]
I'd say one viewer framework is enough burden, whichever one we choose.

**Comment by blink1073** [comment]
@tonysyu, what do you think of a `flask` based version, given your recent experience?

**Comment by stefanv** [comment]
The viewer module has turned out to be very useful so far.  I don't have
anything against it being moved to a separate repo per se, as long as we
keep doing regular releases on it as a team.

**Comment by tonysyu** [comment]
@blink1073: Flask has been pretty good for the toy projects I've been playing around with. I have the (completely unsubstantiated) impression that `tornado` might be preferred for applications where two-way communication is important. (Both matplotlib webagg and IPython notebook use tornado.)

I think I missed something: What's the motivation for rolling our own backend vs using matplotlib's webagg/nbagg? I guess it's not clear how to use widgets with webagg.

**Comment by blink1073** [comment]
@tonysyu, we would not make our own backend, but use either webagg or nbagg.  I'll have to look a bit deeper at tornado vs flask for the application level, as well as what to use for the other widgets (sliders, buttons, etc.).   The simplest would be a web form that updates the image on submit instead of as the user moves the slider, but would not be as interactive/intuitive.  The interactive approach would require websockets.

**Comment by blink1073** [comment]
I created a Wiki [here](https://github.com/scikit-image/scikit-image/wiki/Web-based-Viewer-Proposal) for discussion.

**Comment by tonysyu** [comment]
Hmm, I wish there was support for comments on github wikis. In any case, how did you end up choosing tornado over flask? I've added a couple of questions to the wiki, too.

**Comment by tacaswell** [comment]
Maybe a better way to do this sort of thing is to throw them in the sphinx docs.  You still get reasonable display of the raw rst files from git hub, get comments, and get in your repo history the evolution of the discussion.

**Comment by blink1073** [comment]
Tornado is already used by IPython notebook and the matplotlib web backends, and natively supports websockets.  @tacaswell, there is a history for wikis, and doesn't require a PR to contribute to discussion.

**Comment by blink1073** [comment]
But it would be nice if the wiki supported comments like gists do.

**Comment by hugadams** [comment]
This is a bit offtopic, but since I have all the experts here:

We're looking to create a viewer in which the user can probe around an image OR contour plot with the cursor, and see the x, y and z coordinates, where z is the value of the 2d function at that point.  Looking through mpld3, I don't think there's a widget to return the z value.  I haven't had a chance to explore any of the interactive matplotlib plots that @blink1073 has been putting together.  

Do anyone know of something like this being available, and if not, where it would be possible to implement with the least resistance on our end?

**Comment by tonysyu** [comment]
Hey Adam, I'm not quite sure if this is what you're asking, but if you just want to grab the z value under the cursor, you can just connect an event that grabs the data coordinates and indexes into the image/contour plot. For example:

```python
import matplotlib.pyplot as plt
from skimage import data


fig, ax = plt.subplots()

image = data.chelsea()
ax.imshow(image)

def on_event(event):
    if event.xdata is None or event.ydata is None:
        return

    i_row = int(event.ydata)
    i_col = int(event.xdata)
    print image[i_row, i_col]

fig.canvas.mpl_connect('motion_notify_event', on_event)
plt.show()
```

**Comment by hugadams** [comment]
Awesome.  I'm way behind the curve on interactive MPL.  How long has this behavior been around, and is there a centralized place for MPL interactive examples?   Right now, we're using mpld3 for interactivity in our IPython widget notebooks, and I'm trying to get a handle (lol) on the current an near-future capabilities of matplotlib's interactivity.

**Comment by tacaswell** [comment]
This functionality as been there for a very long time, it is how the updating x, y coordinate in the corner of the windows works.

Also see the mpldatacursor project and https://github.com/matplotlib/matplotlib/pull/3989

**Comment by blink1073** [comment]
@hugadams, you can also put the z-value directly in the canvas by using this technique:  http://stackoverflow.com/a/14755358/3208463.  By interactive, do you mean interactive in the browser/notebook, or interactive in general?  The web/nbagg I'd say are fully featured as of the newest release 1.4.3, meaning they will support anything you can do in Qt, including all of the canvas callbacks.  As far as examples of event handling go, I'd suggest: http://matplotlib.org/examples/event_handling/index.html.  Our own canvas tools make extensive use of the event handling API.  You can see them in action by checking out some of our [viewer examples](https://github.com/scikit-image/scikit-image/tree/master/viewer_examples).

**Comment by blink1073** [comment]
Darn it @tacaswell, I'm busy crediting your Stack Overflow answer and you steal my thunder!

**Comment by hugadams** [comment]
Ok guys, thanks for clearing it up for.  I used MPLD3 originally in the ipython gui video that was linked at the beginning of this thread in part because I thought that was the best way to achieve interactivity in matplotlib plots.  But considering what I've seen here, using MPLD3 directly might be what I want.  For example, I want to create a viewer where the user scrolls a mouse over an image plot (as you Tony has shown), and a projection of the data is drawn in an adjacent plot in realtime.  Something akin to a 2d function inspector (actually here's a Chaco example that's deadon: http://docs.enthought.com/chaco/user_manual/annotated_examples.html#scalar-image-function-inspector-py).  It should be straightforward to trigger redraws as the user scrolls, or even clicks around, yes?  

Another common task I use MPLD3 for a lot is to interactively inspect scatter plots, where clicking on a  scatter point tells me some additional info.  Could I drop MPLD3 and do this within the confines of interactive MPL?  

Sorry to take the thread off into a tangent.

**Comment by stefanv** [comment]
Does anyone know what the most light-weight Javascript enabled browser
widget is?  Because we could then still provide "loose standing"
visualizations (as we currently do), even though we are changing the
underlying infrastructure to web.

**Comment by blink1073** [comment]
@hugadams, you may want to have a look at [image inspector](https://github.com/blink1073/image_inspector), which I think does exactly what you're trying to do.

**Comment by blink1073** [comment]
@stefanv, Html5 has pretty slick `<input>` elements, which you can register jQuery event handlers to.

**Comment by stefanv** [comment]
Right, but what can we use as a lightweight client?  Does the QT widget
support Javascript?

**Comment by blink1073** [comment]
@stefanv, You can put HTML in a Qt widget, but I thought the whole point was to move away from Qt?
I'd say the client end would be a Tornado app, copying what Matplotlib is already doing (and already a requirement for using their web backend).

**Comment by blink1073** [comment]
An update on this: with the addition of https://github.com/matplotlib/matplotlib/pull/7411, our existing canvas tools will be in matplotlib itself.  Additionally, https://github.com/matplotlib/jupyter-matplotlib is a full-fledged ipywidget that works in the Jupyter Notebook.  

Once the Painter tool PR is finished, I propose that we port the existing viewer to use ipywidgets.
We can vendor `matplotlib/widgets.py` and use that if the detected matplotlib version is < 2.1.  

I am willing to work on this if we agree it is the right approach.

**Comment by blink1073** [comment]
Whoops, I spoke too soon, we'll also need the `line` tool, I'll add that one after the painter tool is finished.

**Comment by yxdragon** [comment]
@blink1073 Hello, I am yxdragon, who email to you yesterday. I review all the comment above. I wrote ImagePy https://github.com/yxdragon/imagepy, aimd to build a interactive enviroment with smallest dependent, just: numpy, scipy, wx, and shapely(to surport the diffrent shapes selection and there union, buffer operations...).and with some plugins (can wrap any numpy image procession function with para dialog, preview...),and with some tools (rectangle, ellipse, line, curve, freeline, polygon, zoom, move, painter, colorpicker, measure distance, angle...). I am a new man in github, but I will upgrade the project continuely!

**Comment by blink1073** [comment]
Hello @yxdragon!  Thank you so much for imagepy and for engaging with us.  

A couple of points for the wider group: wx [Pheonix](https://github.com/wxWidgets/Phoenix) (which supports Python 3) has unstable releases and appears to be nearing completion.  The last commit was 3 days ago.  

Also, I asked @yxdragon to consider a BSD-style license, with a pointer to Jake's [blog post](http://www.astrobetter.com/blog/2014/03/10/the-whys-and-hows-of-licensing-scientific-code/).

**Comment by yxdragon** [comment]
@blink1073 I will follow the wx Pheonix, and try to use it in ImagePy. But now I must translate some document (manual and develop) in my free time. I am a Chinese with pool english and I have to work in daytime...and I am a freshman in github and opensourse, I have no knowledge about those licenses, I will just do as your advice, take a BSD license now!

**Comment by yxdragon** [comment]
@blink1073 Thanks for your tips, I will keep on going, and anyone want to join this project is welcom!

**Comment by tacaswell** [comment]
@hugadams Have you looked at `mpldatacursor` for the scatter case?  For the profile, see https://github.com/tacaswell/xray-vision/blob/master/xray_vision/backend/mpl/cross_section_2d.py#L253 for a (very badly factored) example.

Also see the examples in http://matplotlib.org/examples/event_handling/index.html

**Comment by tacaswell** [comment]
One other advantage of using the Matplotlib events is that your code will now drop into what ever GUI frame work you want (including nbagg/ipympl).
